pkg/agent/secondarynetwork/podwatch/k8s/podwatch_controller.go

```diff
@@ -0,0 +1,318 @@
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface,
+	podInformer cache.SharedIndexInformer,
+	resyncPeriod time.Duration,
+	nodeName string,
+	podCache cnipodcache.CNIPodInfoStore,
+	cniServer *cniserver.CNIServer) *PodWatchController {
+
+	pwc := PodWatchController{
+		kubeClient:         kubeClient,
+		netAttachDefClient: netAttachDefClient,
+		podInformer:        podInformer,
+		podLister:          corelisters.NewPodLister(podInformer.GetIndexer()),
+		nodeName:           nodeName,
+		podCache:           podCache,
+		cniServer:          cniServer,
+	}
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    pwc.enqueuePod,
+			DeleteFunc: pwc.enqueuePod,
+			UpdateFunc: func(old, cur interface{}) { pwc.enqueuePod(cur) },
+		},
+		resyncPeriod,
+	)
+
+	pwc.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podwatchcontroller")
+	return &pwc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (podWatcher *PodWatchController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	podWatcher.queue.Add(podKey)
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (podWatcher *PodWatchController) handleAddUpdatePod(obj interface{}) error {
+	var ret bool
+	pod := obj.(*corev1.Pod)
+
+	secondaryNetwork, ret := checkForPodSecondaryNetworkAttachement(pod)
+	if !ret {
+		klog.InfoS("Network Attachement Definition doesn't exist for pod", pod.Name, "namespace", pod.Namespace)
+		return nil
+	}
+
+	//get POD/Container information (container id, namespace, etc.,) from CNI info cache (added by CNI Server CmdAdd upon
+	//successful pod and primary network creation).
+	//If container information is not available yet in the CNICache, return error to requeue work
+	podCNIInfo := podWatcher.podCache.GetCNIConfigInfoPerPod(pod.Name, pod.Namespace)
+	if len(podCNIInfo) == 0 {
+		return errors.New("CNI Cache Information not found")
+	} else {
+		networklist, err := parsePodSecondaryNetworkAnnotation(secondaryNetwork)
+		if err != nil {
+			return errors.New("Secondary network annotation parsing failed")
+		} else {
+			for _, netinfo := range networklist {
+				klog.InfoS("Secondary Network Information:", netinfo)
+				if len(netinfo.NetworkName) > 0 {
+					netDefCRD, err := podWatcher.netAttachDefClient.NetworkAttachmentDefinitions(pod.Namespace).Get(context.TODO(), netinfo.NetworkName, metav1.GetOptions{})
```

Comment:
We do not support changes of NetworkAttachmentDefinition, or NetworkAttachmentDefinition created after Pods?